### PR TITLE
IMAX effect integration

### PR DESCRIPTION
**Pull request by lowderchris**
I've migrated some of the IMAX effect code from punchbowl to here in solpolpy directly. A punchbowl branch exists utilizing these changes, which can be pulled in once this pull request is resolved. Definitely let me know of any changes I can make here to improve. See any TODOs for outstanding questions I can wrap up.

- [x] mark imax effect as done in readme.md

### CODE REVIEW

**Comment by jmbhughes** [comment]
@s0larish could you take a look at the TODOs @lowderchris left in the code? We should make sure we have answers to those before merging.

File: solpolpy/core.py [hunk 1]
```diff
@@ -36,6 +38,9 @@ def resolve(input_data: t.Union[t.List[str], NDCollection], out_system: str) ->
         - "fourpol": For observations taken at sequence of four polarizer angles, i.e. 0°, 45°, 90° and 135°.
         - "npol": Set of images taken at than three polarizing angles other than MZP
 
+    imax_effect : Boolean
```
Comment (jmbhughes) [inline solpolpy/core.py:41]: We should add something in the docs about what the imax effect is. You might link to something in the docs webpages from here for a full explanation.

File: solpolpy/core.py [hunk 2]
```diff
@@ -66,6 +71,11 @@ def resolve(input_data: t.Union[t.List[str], NDCollection], out_system: str) ->
     equation = get_transform_equation(transform_path)
     requires_alpha = check_alpha_requirement(transform_path)
 
+    # TODO - Does this apply for all transformations, or only within MZP->MZP correction?
+    # TODO - If so, should this result in an error when trying to use it outside of this scope?
+    if imax_effect:
```
Comment (jmbhughes) [inline solpolpy/core.py:76]: @s0larish should comment on this. I would like to see an error as you propose if it's specific to MZP->MZP.

File: solpolpy/core.py [hunk 3]
```diff
@@ -158,6 +168,63 @@ def check_alpha_requirement(path: t.List[str]) -> bool:
...
+        for j in range(3):
+            mat_a[:, :, i, j] = (4 * np.cos(phi[i] - thmzp[j]) ** 2 - 1) / 3
+
+    return mat_a
+
+
+def resolve_imax_effect(input_data: NDCollection) -> NDCollection:
+
+    for i, key in enumerate(input_data.keys()):
+        if i == 0:
```
Comment (jmbhughes) [inline solpolpy/core.py:212]: It's not immediately obvious to me what's happening. Please add a comment on why you're checking only the first key. Also, check that order doesn't matter. A dictionary does not guarantee to provide .keys() in a specific order.

File: solpolpy/core.py [hunk 4]
```diff
@@ -158,6 +168,63 @@ def check_alpha_requirement(path: t.List[str]) -> bool:
...
+            data_shape = input_data[key].data.shape
+            data_mzp_camera = np.zeros([data_shape[0], data_shape[1], 3, 1])
+        data_mzp_camera[:, :, i, 0] = input_data[key].data
+
+    imax_matrix = generate_imax_matrix(data_shape)
+    imax_matrix_inv = np.linalg.inv(imax_matrix)
```
Comment (jmbhughes) [inline solpolpy/core.py:218]: Will this inversion ever fail?

**Review by jmbhughes — CHANGES_REQUESTED** [review]
I would like to see more tests. Is there a simple test case for IMAX matrix we can generate to actually check the values? We need to make sure the matrix is properly constructed. Or could we make sure its output is different when its applied than not applied. It would be particularly nice if the ratio of output with and without the IMAX effect has a nice property we can check for. I'm just brainstorming some tests here.

File: solpolpy/core.py (hunk 4 above)
Comment (s0larish) [reply solpolpy/core.py:218]: For singular matrix case there is a chance. However it didn't fail yet for the WFI cases we tested. Even with some weirdest angles we passed as input (like 90 degrees)

File: solpolpy/core.py (hunk 1 above)
Comment (s0larish) [reply solpolpy/core.py:41]: We can refer our paper once it's out. Till then we have some poster (PUNCH 4) and abstract (AGU) to refer to.

File: solpolpy/core.py (hunk 2 above)
Comment (s0larish) [reply solpolpy/core.py:76]: Yes it's specific to MZP -> MZP. We can make an estimate if this correction is not done on WFI what percentage of error in pB we can get.

File: solpolpy/core.py (hunk 4 above)
Comment (jmbhughes) [reply solpolpy/core.py:218]: https://numpy.org/doc/stable/reference/generated/numpy.linalg.LinAlgError.html

File: tests/test_core.py [hunk 5]
```diff
@@ -260,3 +270,17 @@ def test_btbr_to_mzp(btbr_ones):
 def test_bp3_to_bthp(bp3_ones):
     result = resolve(bp3_ones, "Bthp")
     assert isinstance(result, NDCollection)
+
+
+def test_imax_effect(mzp_data):
+    result = resolve(mzp_data, "MZP", imax_effect=True)
+    assert isinstance(result, NDCollection)
+    for key in result.keys():
+        print(np.sum(result[key].data * mzp_data[key].data))
```
Comment (jmbhughes) [inline tests/test_core.py:279]: Please remove the print from the test.

**Review by jmbhughes — CHANGES_REQUESTED** [review]
Almost there. These should be easy things to fix I think.

File: solpolpy/core.py (hunk 3 above)
Comment (jmbhughes) [reply solpolpy/core.py]: Still need to clarify/resolve this section.

**Review by jmbhughes — APPROVED** [review]

**Pull request merged**